semantic-tags/src/Tags/Tagging/Precise.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE AllowAmbiguousTypes, ConstraintKinds, FlexibleContexts, FlexibleInstances, MultiParamTypeClasses, RankNTypes, ScopedTypeVariables, TypeApplications, TypeOperators #-}
+module Tags.Tagging.Precise
+( Tags
+, ToTags(..)
+, yield
+, runTagging
+, GFoldable1(..)
+) where
+
+import Control.Effect.Pure
+import Control.Effect.Reader
+import Control.Effect.Writer
+import Data.Monoid (Endo(..))
+import GHC.Generics
+import Prelude hiding (span)
+import Source.Loc (Loc)
+import Source.Span
+import Source.Source
+import Tags.Tag
+
+type Tags = Endo [Tag]
+
+class ToTags t where
+  tags :: Source -> t Loc -> [Tag]
+
+
+yield :: (Carrier sig m, Member (Writer Tags) sig) => Tag -> m ()
+yield = tell . Endo . (:) . modSpan toOneIndexed where
+  modSpan f t@Tag{ span = s } = t { span = f s }
+  toOneIndexed (Span (Pos l1 c1) (Pos l2 c2)) = Span (Pos (l1 + 1) (c1 + 1)) (Pos (l2 + 1) (c2 + 1))
+
+runTagging :: Source -> ReaderC Source (WriterC Tags PureC) () -> [Tag]
+runTagging source
+  = ($ [])
+  . appEndo
+  . run
+  . execWriter
+  . runReader source
+
+
+-- FIXME: move GFoldable1 into semantic-ast.
+class GFoldable1 c t where
+  -- | Generically map functions over fields of kind @* -> *@, monoidally combining the results.
+  gfoldMap1
+    :: Monoid b
+    => (forall f . c f => f a -> b)
```

Comment:
This is _incredibly_ useful for running actions over every subterm position in a syntax tree.

A slight generalization would also take a function `(a -> b)` to apply to the `Par1`s (occurrences of the type parameter `a`, i.e. the annotations), but I didn’t need it for this, so I left it for now.